bigtable-client-core-parent/bigtable-client-core/src/main/java/com/google/cloud/bigtable/config/BigtableOptions.java

```diff
@@ -15,51 +15,103 @@
 
 /**
  * An immutable class providing access to configuration options for Bigtable.
  *
  * @author sduskis
  * @version $Id: $Id
  */
-//TODO: Perhaps break this down into smaller options objects?
-//TODO: This should be @Autovalue + Builder
-public class BigtableOptions implements Serializable, Cloneable {
+@AutoValue
+public abstract class BigtableOptions implements Serializable, Cloneable {
 
   private static final long serialVersionUID = 1L;
 
-  // If set to a host:port address, this environment variable will configure the client to connect
-  // to a Bigtable emulator running at the given address with plaintext negotiation.
+  // If set to a host:port address, this environment variable will configure the
+  // client to connect
+  // to a Bigtable emulator running at the given address with plaintext
+  // negotiation.
   // TODO: Link to emulator documentation when available.
-  /** Constant <code>BIGTABLE_EMULATOR_HOST_ENV_VAR="bigtableadmin.googleapis.com"</code> */
+  /**
+   * Constant
+   * <code>BIGTABLE_EMULATOR_HOST_ENV_VAR="bigtableadmin.googleapis.com"</code>
+   */
   public static final String BIGTABLE_EMULATOR_HOST_ENV_VAR = "BIGTABLE_EMULATOR_HOST";
 
-  /** Constant <code>BIGTABLE_ADMIN_HOST_DEFAULT="bigtableadmin.googleapis.com"</code> */
-  public static final String BIGTABLE_ADMIN_HOST_DEFAULT =
-      "bigtableadmin.googleapis.com";
-  /** Constant <code>BIGTABLE_DATA_HOST_DEFAULT="bigtable.googleapis.com"</code> */
+  /**
+   * Constant
+   * <code>BIGTABLE_ADMIN_HOST_DEFAULT="bigtableadmin.googleapis.com"</code>
+   */
+  public static final String BIGTABLE_ADMIN_HOST_DEFAULT = "bigtableadmin.googleapis.com";
+  /**
+   * Constant <code>BIGTABLE_DATA_HOST_DEFAULT="bigtable.googleapis.com"</code>
+   */
   public static final String BIGTABLE_DATA_HOST_DEFAULT = "bigtable.googleapis.com";
-  /** Constant <code>BIGTABLE_BATCH_DATA_HOST_DEFAULT="bigtable.googleapis.com"</code> */
+  /**
+   * Constant
+   * <code>BIGTABLE_BATCH_DATA_HOST_DEFAULT="bigtable.googleapis.com"</code>
+   */
   public static final String BIGTABLE_BATCH_DATA_HOST_DEFAULT = "batch-bigtable.googleapis.com";
   /** Constant <code>BIGTABLE_PORT_DEFAULT=443</code> */
   public static final int BIGTABLE_PORT_DEFAULT = 443;
 
-  /** Constant <code>BIGTABLE_DATA_CHANNEL_COUNT_DEFAULT=getDefaultDataChannelCount()</code> */
+  /**
+   * Constant
+   * <code>BIGTABLE_DATA_CHANNEL_COUNT_DEFAULT=getDefaultDataChannelCount()</code>
+   */
   public static final int BIGTABLE_DATA_CHANNEL_COUNT_DEFAULT = getDefaultDataChannelCount();
 
-  /** Constant <code>BIGTABLE_APP_PROFILE_DEFAULT=""</code>, defaults to the server default app profile */
+  /**
+   * Constant <code>BIGTABLE_APP_PROFILE_DEFAULT=""</code>, defaults to the
+   * server default app profile
+   */
   public static final String BIGTABLE_APP_PROFILE_DEFAULT = "";
 
   private static final Logger LOG = new Logger(BigtableOptions.class);
+  @Nullable
+  public abstract String appProfileId();
```

Comment:
Is there a way to keep these methods as `setXXX()` instead of `xXX`?

Reply:
We do have setters with name setXXX() like setAppProfileId().

Reply:
Sorry... I should have said getters :).  We currently have `getAppProfileId()` and changing the getter would break our users.